### DISCUSSION TITLE
Making plugins truly multilingual

### DISCUSSION THREAD

**Discussion by NicolasPetton** [TiddlyWiki/TiddlyWiki5#5393] [Ideas]
Hi!

I'm currently in the process of adding support for additional languages to [Projectify](https://projectify.wiki). I'm following the same structure and patterns as the core language plugins, which works fine. 

However I cannot (or I've missed it) **extend** existing languages (say fr-FR) so that Projectify uses the language of the wiki. This means that the user has to enable/disable Projectify language plugins to choose the language for Projectify, independently from the the language of the core, which is poor UX.

I might have missed something, but if that's not the case, I'd like to contribute improvements to language support in plugins to improve the user experience for switching languages.

Cheers,
Nico

**Comment by pmario** [comment]
Hi, 
At the moment it's hard to create different languages for plugins. The biggest problem I see, is that TW has 20+ translations for the core elements. As a plugin author you will not be able to provide all those languages with your plugin. 

and

The biggest disadvantage will be pugin size. Many of my plugins are relatively small. So I do avoid to add the full license text. I only use links to the text, because the license text will more weight than I want. 

I do have a private version of TW, that stripped out the core language and uses German instead, because I don't want to carry the extra bytes. ...

**Comment by pmario** [comment]
We can install new languages using the default TW plugin mechanism. ... As a plugin author it should be possible to create your own language-addOns ... that can be loaded from your library list. ... 

So users would be able to install only those texts, that they want. .. 

The disadvantage: You would have 1 plugin and worst case 20+ language plugins .. So if you have 21 plugins you'd probably end up with 400+ plugins, just to manange your texts. ... IMO that's a mess

**Reply by NicolasPetton** [reply]
> The disadvantage: You would have 1 plugin and worst case 20+ language plugins .. So if you have 21 plugins you'd probably end up with 400+ plugins, just to manange your texts. ... IMO that's a mess

To me the biggest UX issue is that language-addons are not in sync with the language of the wiki. If I switch my wiki to Spanish, none of the plugins will switch to Spanish.

**Reply by pmario** [reply]
I know. I did create a "user story" at: https://github.com/Jermolene/TiddlyWiki5/discussions/5399#discussioncomment-279274

**Reply by pmario** [reply]
I would like to be able to create a "plugin language pack", that contains all of my additionally needed texts for all of my plugins. .. Or may be most of my plugins. .. It depends on the plugin. 

At the moment I try to limit my texts to core texts, that match. .. BUT most of the time I use "hard coded" english

**Reply by silvyn** [reply]
> The disadvantage: You would have 1 plugin and worst case 20+ language plugins .. So if you have 21 plugins you'd probably end up with 400+ plugins, just to manange your texts. ... IMO that's a mess

IMO it's a user choice !
If Projectify is translate to 20+ language and more, good for him !

But I doubt that it can be done in 1 day, and if there are already 4/5 languages, it will already be a big step forward!

I'd like to be able to translate plugins, and I think TW should authorise/propose a method to facilitate it.

Otherwise having to overwrite the plugins and find out where the texts to be changed are, it's almost mission impossible... and that's a shame. Not everyone speaks English. And that complicates the task if there are updates.

**Comment by pmario** [comment]
@NicolasPetton 

I think I found a possibility, that can be easily supported by the core. ... TW already has a "language-switcher". The switcher is automatically activated if the system tiddler `$:/language` is changed. 

It's relatively easy to add a second handler, that can handle `plugin-type: language-addon`. Those language-addons work in the exact same way as the core language tiddlers. 

BUT .. They are initialized after the main `language` tiddlers. They can create new language elements or overwrite existing ones. ... 

I'll create an experimental PR soon. So you should be able to test it.

**Reply by NicolasPetton** [reply]
Hi @pmario,

I really like your approach, I'd be happy to test this in an experimental branch.

Cheers,
Nico

**Comment by Jermolene** [comment]
Hi @NicolasPetton we do indeed lack proper support for making plugins translatable, I do think it's important to address. 

One point to note about the language plugin architecture is that the core goes to elaborate lengths to switch language plugins on and off in their entirety when the core language is switched. In other words, if one has several language plugins loaded into a wiki, only the currently selected one will be active, with the shadow tiddlers from the other plugins being hidden. I now think that that approach was a mistake: it would be sufficient to have had `$:/language` contain the language code to be used to retrieve language strings, instead of the name of a single plugin.

Anyhow, @pmario's proposed `plugin-type: language-addon` sounds like the right approach. It's worth noting that the same implementation could also be made to work with themes, so that people could make plugins that work as add-ons to a theme.

**Reply by NicolasPetton** [reply]
> Anyhow, @pmario's proposed plugin-type: language-addon sounds like the right approach. 

I think so too.

> It's worth noting that the same implementation could also be made to work with themes, so that people could make plugins that work as add-ons to a theme.

That would be great, and I can already see a few personal use-cases for theme add-ons.

**Comment by pmario** [comment]
>...I now think that that approach was a mistake: it would be sufficient to have had $:/language contain the language code to be used to retrieve language strings, instead of the name of a single plugin.

I don't think so. TW already has 2000+ shadow tiddlers. I think, they are not indexed. .. So if you have 3 languages installed, this number is still the same. IMO that's a good thing.

If all of them would be active at the same time, we would have 6000 shadows, for no big win, ... just bloat.

As I wrote, It's possible to mount a language into the core directory and compile a version that doesn't contain the "fallback" language strings but is smaller!

**Comment by pmario** [comment]
@NicolasPetton see: https://github.com/Jermolene/TiddlyWiki5/pull/5418 @Jermolene

**Comment by pmario** [comment]
@NicolasPetton I did upload a video, where you can see, how the structure _can_ look like. https://youtu.be/yZMaZ4qeaKk

I did test the workflow for TW core plugins _and_ 1 of my 3rd party plugins. It works equally well. .. The UI needs to be "simplified"

**Reply by NicolasPetton** [reply]
Thanks Mario, I'll check out the branch to test it locally.

**Reply by pmario** [reply]
hi, Make sure your environment variables are set in the right way. So the library build process can find them. You dont need to copy your plugins into the TW file structure. see: https://tiddlywiki.com/#Environment%20Variables%20on%20Node.js

**Reply by NicolasPetton** [reply]
I know, but thanks anyway! :)
